### PR TITLE
Add "Array" base class, which provides numpy-style indexing.

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

Augment `StreamingDataset`, `LocalDataset`, and `Reader` (i.e., shards) with fancy numpy-style indexing. This is done by providing an implementation of `__getitem__`. Subclasses should now override `get_item(0 <= idx < size)`, which is called by our `__getitem__` for individual samples.

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Review by dakinggg — COMMENTED** [review]
broadly LGTM. One concern is that this is a breaking change, right? e.g. currently in the examples repo the dataset implements __getitem__, and with this change it would need to implement getitem instead?

**Comment by dakinggg** [comment]
can you add `slice(50, -1000, -5)` test case? It gives different output for numpy/yours for me locally.

**Comment by knighton** [comment]
Yep, `examples` would need to be updated to get the new functionality. If not updated, it would continue working as 0 <= idx < N as before

**Comment by dakinggg** [comment]
> Yep, `examples` would need to be updated to get the new functionality. If not updated, it would continue working as 0 <= idx < N as before

I think it would crash if not updated, because it calls `super().__getitem__()`, which in turn calls `self.getitem()`, which would not be implemented.

**Comment by knighton** [comment]
Oof! Appear to have forgotten a check. Thanks

**Comment by knighton** [comment]
> can you add slice(50, -1000, -5) test case? It gives different output for numpy/yours for me locally.

Fixed :1000:

File: tests/test_array.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    def size(self) -> int:
+        return self.num_items
+
+    def get_item(self, idx: int) -> int:
+        assert 0 <= idx < self.num_items
+        return idx
+
+
+size = 100
+theirs = np.arange(size)
```
[the remark this answers is not in the document]
Comment (knighton) [reply tests/test_array.py]: done

[the remark this answers is not in the document]
Comment (knighton) [reply tests/test_array.py]: done

**Review by dakinggg — APPROVED** [review]
LGTM

**Comment by LangDaoAI** [comment]
Great feature!  when to merge into next release?  thanks!

**Review by karan6181 — COMMENTED** [review]
Overall looks excellent. Few comments:

- please also change the getitem() method for https://github.com/mosaicml/streaming/blob/main/streaming/multimodal/webvid.py#L39.
- Have you ran any end-to-end integration test from example repo such as BERT / LLM to ensure this functionality doesn't break anything?

**Comment by knighton** [comment]
LGTM

**Pull request merged**